### PR TITLE
Allow to disable HTTP(S) in Debian

### PR DESCRIPTION
**Pull request by t2d**
I think the best way to do it, is to include a (OS-specific) template.

### CODE REVIEW

**Comment by geerlingguy** [comment]
I would want to do this in a way that works with both RHEL and Debian. Not opposed to this idea in general, but it seems like it could be a little fragile, and it also excludes using any other kinds of ports configurations (managed outside this role).

But I haven't done a full review, just a quick glance on my mobile of the changes.

**Comment by oxyc** [comment]
https://github.com/geerlingguy/drupal-vm/issues/487#issue-137647691 also wanted the possibility to use different ports per vhost. This is supported in the nginx role so maybe it's worth adding here as well.

**Comment by oxyc** [comment]
I had a look at getting this working for RHEL but it does look really difficult. Here are some of my thoughts.

The current related issues are:
- Cannot listen to more than one port #43
- Cannot specify the port on a per vhost basis https://github.com/geerlingguy/drupal-vm/issues/487
- Cannot disable listening for SSL #30
- Not setting NameVirtualHost causes SSL issues #21, https://github.com/geerlingguy/drupal-vm/issues/321

The take the fact that we should support modifications by other roles. This is very fragile at the moment as we search for all listen directives and just replace them.

One solution would be to use `replace` module and replace bundles of Listen directives. For example something like:
```yaml
- regexp: "^(Listen [0-9]+\n+)+"
  replace: "Listen {{ apache_listen_port.split('\nListen ') }}\n"
```
This is very easy to break by a third party role though. Which is the current situation as well.

In my opinion the requirement for third party roles to manipulate the configuration files should not be supported. If we want to solve all the issues related to ports I really see no other option than doing what @t2d does and doing the same for RHELs `httpd.conf` and `ssl.conf` as well. To allow modifications we could expose the template file and let people override it.

Even this solution sounds quite unmaintainable as we support: RHEL, Suse, Solaris as well as Debian and two versions of Apache. That's potentially an insane amount config templates...

**Comment by t2d** [comment]
> Even this solution sounds quite unmaintainable as we support: RedHat, Suse, Solaris as well as Debian and two versions of Apache. That's potentially an insane amount config templates...

Well, for Debian and Apache 2.2+2.4 it was one file. And the structure is way easier to understand than before.
Lineinfile is a great module and I try to use it all the time. But as you say, this role is quite complex. This role is the general go-to solution for Apache with ansible. Anyone can write a simple OS specific role. I personally see no need in supporting double handling of Apache.

**Comment by oxyc** [comment]
Unfortunately RedHat requires changes in two `httpd.conf` and `ssl.conf`. I'm also guessing these look different depending on Apache 2.2 vs 2.4. Solaris and Suse I'm not sure. I guess it's not so bad if Solaris and Suse can use the same templates as Debian or RedHat.

**Comment by oxyc** [comment]
Another option that might work better is include a `ports.conf` debian-style config that we add an include directive for in `httpd.conf`. With that we can just iterate over all ports and delete lines beginning with `Listen {{ port }}` within `httpd.conf`. That ensures we only modify the ones we're responsible of.

**Comment by geerlingguy** [comment]
Yeah... and in some ways, the whole thing is a non-starter for me, just because it seems Apache is not built to easily handle this idea of ports-per-vhost concept like Nginx's port-per-server configuration... I don't know if I want to add on the complexity/baggage for what is really a 20% use case (probably far less than that, based on what I've seen in real world usage.

Almost anyone with requirements this specific decides to go with Nginx anyways, because of the per-server/host flexibility that it affords out of the box.

**Comment by geerlingguy** [comment]
Looking at @oxyc's suggestion again...

> Another option that might work better is include a ports.conf debian-style config that we add an include directive for in httpd.conf. With that we can just iterate over all ports and delete lines beginning with Listen {{ port }} within httpd.conf. That ensures we only modify the ones we're responsible for.

I think I'd be fine going this route. I would like to make sure we test the changed config to make sure things don't break... but this looks acceptable.

**Comment by choonming** [comment]
@geerlingguy any chance for this to be merged or similar solutions for this?

File: defaults/main.yml [hunk 1]
```diff
@@ -3,6 +3,9 @@ apache_enablerepo: ""
 
 apache_listen_port: 80
 apache_listen_port_ssl: 443
+apache_listen_ports:
```
Comment (geerlingguy) [inline defaults/main.yml]: If we're adding this new variable, it needs to be documented in the README.

File: vars/Debian.yml [hunk 2]
```diff
@@ -8,7 +8,3 @@ apache_conf_path: /etc/apache2
 __apache_packages:
   - apache2
   - apache2-utils
-
-apache_ports_configuration_items:
```
Comment (geerlingguy) [inline vars/Debian.yml]: Would we also need to make a change in the RedHat config?

**Comment by t2d** [comment]
I'm not using this role anymore. Feel free to close the PR or work on top of it. Best regards

**Comment by ByJacob** [comment]
For the future - if you want add second port in Debian:
1. add task
```yaml
- name: Add second apache2 ports
      lineinfile:
        dest: "/etc/apache2/ports.conf"
        state: present
        regexp: "^[ ]Listen 8080"
        line: " Listen 8080"
```
2. apache config
```yaml
apache_global_vhost_settings: |
  DirectoryIndex index.php index.html
  <VirtualHost 127.0.0.1:8080>

    DocumentRoot /home/www/{{instance_name}}/backend/web
```

**Pull request closed without merging**